### PR TITLE
client: anchor dentries for trimming to make cap traversal safe

### PR DESCRIPTION
**Pull request by batrick**
Fixes: http://tracker.ceph.com/issues/24137

Signed-off-by: Patrick Donnelly <pdonnell@redhat.com>

### CODE REVIEW

**Comment by batrick** [comment]
@ukernel thanks for the reproducer. I'm working on adding it to a QA test.

I think this fix is simpler than what you had posted in [1]; I don't like the absurd complexity of deleting dentries during traversal, so let's not do it.

[1] http://tracker.ceph.com/issues/24137

File: src/client/Client.cc [hunk 1]
```diff
@@ -4138,7 +4137,10 @@ void Client::trim_caps(MetaSession *s, uint64_t max)
       }
     }
   }
-  ldout(cct, 20) << " clearing anchored inodes" << dendl;
+  ldout(cct, 20) << " clearing anchored dentries" << dendl;
```
Comment (ukernel) [inline src/client/Client.cc]: "trimming anchored dentries"

File: src/client/Client.cc [hunk 2]
```diff
@@ -4089,7 +4089,7 @@ void Client::trim_caps(MetaSession *s, uint64_t max)
 
   uint64_t trimmed = 0;
   auto p = s->caps.begin();
-  std::set<InodeRef> anchor; /* prevent put_inode from deleting all caps during traversal */
+  std::set<Dentry *> anchor; /* prevent put_inode from deleting all caps during traversal */
```
Comment (ukernel) [inline src/client/Client.cc]: maybe rename 'anchor' to 'to_trim'

File: src/client/Client.cc (hunk 1 above)
Comment (batrick) [reply src/client/Client.cc]: done, thanks

File: src/client/Client.cc (hunk 2 above)
Comment (batrick) [reply src/client/Client.cc]: done, thanks

**Comment by batrick** [comment]
http://pulpito.ceph.com/pdonnell-2018-05-18_23:51:38-fs:bugs-wip-pdonnell-i24137-distro-basic-smithi/

for 63fe15725740b696f030ccc5408f6d9db2532c09

**Comment by ukernel** [comment]
I think it's better to not add the qa test.  It's too tricky to arrange inodes in session->caps. we don't know if it still works in the future.

**Comment by batrick** [comment]
It doesn't hurt to run the test so I'm inclined to keep it for now. It serves as a useful template for other tests. I'll add a note that it's unlikely to continue testing as expected.

**Pull request merged**